### PR TITLE
Fix `sincos` for reals that are not AbstractFloats

### LINKED ISSUE #22095: sincos is broken for nonstandard types
Due to 
https://github.com/JuliaLang/julia/blob/3e3f96630cd9ad095c0b12b604e8d3d666b3c34b/base/math.jl#L428:

```jl
 res = Base.FastMath.sincos_fast(x)
```
This breaks `sin`, `cos`, `exp` etc. for `Complex{T}` with non-standard `T` (i.e. not `Float64`, `Float32`), that previously worked fine (but apparently weren't tested).

Presumably it's sufficient to restrict the argument type of the above `sincos` definition and add a generic fallback? Should that be

```jl
sincos(x) = (sin(x), cos(x))
```
or is there a better solution?

### PR DESCRIPTION
**Pull request by dpsanders**
Fixes https://github.com/JuliaLang/julia/issues/22095.

This PR adds a generic fallback for `sincos`.
This allows `exp(a + im*b)` to work correctly when the type of `a` and `b` is a non-standard subtype of `Real` (e.g. `ArbFloat` or `Interval`). 

There is a test provided with the simplest such non-standard type.

[Note that this -- EDIT: that is, `exp(a + im*b)` -- worked in Julia 0.6, but does not work in current master, so this is in fact a bug fix for a regression and not a new feature.]

### CODE REVIEW

File: base/special/trig.jl [hunk 1]
```diff
@@ -197,7 +197,9 @@ function sincos(x::T) where T<:Union{Float32, Float64}
         return -co, si
     end
 end
-sincos(x::Real) = sincos(float(x))
+sincos(x::T) where {T <: Union{Integer, Rational}} = sincos(float(x))
```
Comment (giordano) [inline base/special/trig.jl]: What about irrational?

Comment (giordano) [reply base/special/trig.jl]: Yes, but irrationals don't use the faster version, right? (I'm on my phone, I can't check myself right now)

Comment (pkofod) [reply base/special/trig.jl]: Oh, my bad, of course you're right, it should be `{Integer, Rational, Irrational}`. This changes it such that `sincos(pi)` is now slower than before.

Comment (dpsanders) [reply base/special/trig.jl]: Ah, good point, thanks, I'll change it.

I don't understand how to interpret the test errors -- can anybody help me?

**Comment by fredrikekre** [comment]
`LoadError("sysimg.jl", 320, LoadError("math.jl", 988, LoadError("special/trig.jl", 201, UndefVarError(:Irrational))))`

**Comment by pkofod** [comment]
Is irrational not defined at this point?

**Comment by dpsanders** [comment]
Apparently not. Also, in the future we may want to define sin(pi) etc explicitly. So I think it's better to leave the generic definition to cover irrationals.

**Comment by pkofod** [comment]
Practically speaking this should be of no real importance, but do you then propose that sincos(someotherrational) should be slow? Or do you want to define sin for all irrationals?

**Comment by dpsanders** [comment]
I have now implemented @yuyichao's approach from https://github.com/JuliaLang/julia/issues/22095#issuecomment-304417064

and left a generic fallback.
I think this is the best solution for now.

**Review by stevengj — APPROVED** [review]

**Comment by dpsanders** [comment]
The current failures seem unrelated (being in the `show` tests, not the `math` tests).

**Comment by fredrikekre** [comment]
It is related
```
Error in testset show:
Test Failed at /tmp/julia/share/julia/test/show.jl:580
  Expression: contains(Base.url(first(methods(sin))), "https://github.com/JuliaLang/julia/tree/$((Base.GIT_VERSION_INFO).commit)/base/missing.jl#L")
```

**Comment by JeffBezanson** [comment]
That is a very unfortunately specific test. It should be revised or removed so this can be merged.

**Comment by dpsanders** [comment]
I have now reimplemented the logic using multiple dispatch instead.
I fixed the overly-specific test following Jameson's suggestion.

Not sure what the test failures are about now. They seem unrelated?

**Comment by dpsanders** [comment]
Made a big rebase/merge mess again, sorry. How do I back myself out of this one?

**Comment by dpsanders** [comment]
Have now fixed the mess by cherry-picking onto latest master and force pushing.

**Comment by dpsanders** [comment]
Are the failures in Circle x86_64 related? If not, then this is ready to go.

**Review by stevengj — APPROVED** [review]

**Pull request merged**